File: hydra-python/src/main/python/hydra/lib/eithers.py
"""Python implementations of hydra.lib.eithers primitives."""

from __future__ import annotations
from collections.abc import Callable
from typing import TypeVar
from hydra.dsl.python import Either, Left, Right, frozenlist

A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")
D = TypeVar("D")


def bimap(f: Callable[[A], C], g: Callable[[B], D], e: Either[A, B]) -> Either[C, D]:
    """Map over both sides of an Either value."""
    match e:
        case Left(val):
            return Left(f(val))
        case Right(val):
            return Right(g(val))


def either(f: Callable[[A], C], g: Callable[[B], C], e: Either[A, B]) -> C:
    """Eliminate an Either value by applying one of two functions."""
    match e:
        case Left(val):
            return f(val)
        case Right(val):
            return g(val)


def is_left(e: Either[A, B]) -> bool:
    """Check if an Either is a Left value."""
    return isinstance(e, Left)


def is_right(e: Either[A, B]) -> bool:
    """Check if an Either is a Right value."""
    return isinstance(e, Right)


def map_left(f: Callable[[A], C], e: Either[A, B]) -> Either[C, B]:
    """Map a function over the Left side of an Either."""
    match e:
        case Left(val):
            return Left(f(val))
        case Right(val):
            return Right(val)


def map_right(f: Callable[[B], C], e: Either[A, B]) -> Either[A, C]:
    """Map a function over the Right side of an Either."""
    match e:
        case Left(val):
            return Left(val)
        case Right(val):
            return Right(f(val))


def from_left(default: A, e: Either[A, B]) -> A:
    """Extract the Left value, or return a default."""
    match e:
        case Left(val):
            return val
        case Right(_):
            return default


def from_right(default: B, e: Either[A, B]) -> B:
    """Extract the Right value, or return a default."""
    match e:
        case Left(_):
            return default
        case Right(val):
            return val


def lefts(eithers: frozenlist[Either[A, B]]) -> frozenlist[A]:
    """Extract all Left values from a list of Eithers."""
    result: list[A] = []
    for e in eithers:
        match e:
            case Left(val):
                result.append(val)
            case Right(_):
                pass
    return tuple(result)


def rights(eithers: frozenlist[Either[A, B]]) -> frozenlist[B]:
    """Extract all Right values from a list of Eithers."""
    result: list[B] = []
    for e in eithers:
        match e:
            case Left(_):
                pass
            case Right(val):
                result.append(val)
    return tuple(result)


def partition_eithers(eithers: frozenlist[Either[A, B]]) -> tuple[frozenlist[A], frozenlist[B]]:
    """Partition a list of Eithers into lefts and rights."""
    left_vals: list[A] = []
    right_vals: list[B] = []
    for e in eithers:
        match e:
            case Left(val):
                left_vals.append(val)
            case Right(val):
                right_vals.append(val)
    return (tuple(left_vals), tuple(right_vals))

File: hydra-java/src/main/java/hydra/dsl/prims/Eithers.java
package hydra.dsl.prims;

import hydra.core.Term;
import hydra.lib.eithers.Bimap;
import hydra.lib.eithers.Bind;
import hydra.lib.eithers.Either;
import hydra.lib.eithers.FromLeft;
import hydra.lib.eithers.FromRight;
import hydra.lib.eithers.IsLeft;
import hydra.lib.eithers.IsRight;
import hydra.lib.eithers.Lefts;
import hydra.lib.eithers.PartitionEithers;
import hydra.lib.eithers.Rights;

/**
 * DSL interface providing either primitive operations.
 */
public interface Eithers {
    /**
     * Returns a term representing the bimap primitive operation for eithers.
     * Applies two functions to the left and right cases respectively.
     *
     * @return a term for bimap over eithers
     */
    static Term bimap() {
        return new Bimap().term();
    }

    /**
     * Returns a term representing the bind primitive operation for eithers.
     * Monadic bind for the Either type.
     *
     * @return a term for monadic bind in an either context
     */
    static Term bind() {
        return new Bind().term();
    }

    /**
     * Returns a term representing the either primitive operation.
     * Case analysis for the Either type.
     *
     * @return a term for case analysis over eithers
     */
    static Term either() {
        return new Either().term();
    }

    /**
     * Returns a term representing the fromLeft primitive operation.
     * Extracts the value from a Left, using a default if Right.
     *
     * @return a term for extracting the left value
     */
    static Term fromLeft() {
        return new FromLeft().term();
    }

    /**
     * Returns a term representing the fromRight primitive operation.
     * Extracts the value from a Right, using a default if Left.
     *
     * @return a term for extracting the right value
     */
    static Term fromRight() {
        return new FromRight().term();
    }

    /**
     * Returns a term representing the isLeft primitive operation.
     * Tests if an Either is a Left value.
     *
     * @return a term for testing if left
     */
    static Term isLeft() {
        return new IsLeft().term();
    }

    /**
     * Returns a term representing the isRight primitive operation.
     * Tests if an Either is a Right value.
     *
     * @return a term for testing if right
     */
    static Term isRight() {
        return new IsRight().term();
    }

    /**
     * Returns a term representing the lefts primitive operation.
     * Extracts all Left values from a list of Eithers.
     *
     * @return a term for extracting lefts
     */
    static Term lefts() {
        return new Lefts().term();
    }

    /**
     * Returns a term representing the partitionEithers primitive operation.
     * Partitions a list of Eithers into a pair of lists.
     *
     * @return a term for partitioning eithers
     */
    static Term partitionEithers() {
        return new PartitionEithers().term();
    }

    /**
     * Returns a term representing the rights primitive operation.
     * Extracts all Right values from a list of Eithers.
     *
     * @return a term for extracting rights
     */
    static Term rights() {
        return new Rights().term();
    }
}
